colcon_package_selection/package_selection/dependencies.py

```diff
@@ -103,13 +118,23 @@ def check_parameters(self, args, pkg_names):  # noqa: D102
             sys.exit('\n'.join(error_messages))
 
     def select_packages(self, args, decorators):  # noqa: D102
-        if args.packages_up_to:
-            select_pkgs = set(args.packages_up_to)
+        if (
+            args.packages_up_to is not None or
+            args.packages_up_to_regex is not None
+        ):
+            select_pkgs = set(args.packages_up_to or {})
+            patterns = args.packages_up_to_regex or []
             for decorator in reversed(decorators):
-                if decorator.descriptor.name in select_pkgs:
+                pkg = decorator.descriptor
+                if (
+                    pkg.name in select_pkgs or
+                    any(
+                        re.match(pattern, pkg.name)
+                        for pattern in patterns
+                    )
```

Comment:
Implementing the new option merged into the logic of the existing option (at least as it is atm) isn't correct. When multiple different arguments to select packages is used the boolean logic to combines them is: packages must satisfy both conditions (rather than either of them as currently implemented). Maybe keeping the logic for the new option in a separate block is easier?

Reply:
Ah, I overlooked that. I split it out in a73e2997549918b8b62042afa3a0982b8643fd53.

Seems to be working:

```
$ colcon list -n --packages-up-to rclcpp --packages-up-to-regex .*trace
```

doesn't contain `tracetools_test`/`tracetools_analysis`/etc.